src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/QualifiedNameBuilder.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editing;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Extensions.Default
+{
+    /// <summary>
+    /// Use this utility to build a QualifiedName from a string.
+    /// </summary>
+    public static class QualifiedNameBuilder
+    {
+        /// <summary>
+        /// Use this utility to build a QualifiedName from a string.
+        /// </summary>
+        /// <param name="generator">Call SyntaxGenerator.GetGenerator(document) to get a generator that builds this syntax in a language agnostic way.</param>
+        /// <param name="qualifiedName">The string representing qualified syntax (e.g. Microsoft.AspNetCore.Mvc.Controller).</param>
+        /// <returns>A QualifiedSyntaxNode. An IdentifierNameSyntax is returned when there is no concept of "left" and "right" nodes.</returns>
+        public static SyntaxNode BuildQualifiedNameSyntax(SyntaxGenerator generator, string qualifiedName)
+        {
+            if (generator is null)
+            {
+                throw new ArgumentNullException(nameof(generator));
+            }
+
+            if (qualifiedName is null || string.IsNullOrWhiteSpace(qualifiedName))
+            {
+                throw new ArgumentNullException(nameof(qualifiedName));
+            }
+
+            return BuildRecursiveNameSyntax(generator, qualifiedName.Split('.'), generator.QualifiedName);
+        }
+
+        /// <summary>
+        /// Use this utility to build a MemberAccessExpression from a string.
+        /// </summary>
+        /// <param name="generator">Call SyntaxGenerator.GetGenerator(document) to get a generator that builds this syntax in a language agnostic way.</param>
+        /// <param name="qualifiedName">The string representing a member access expression (e.g. Microsoft.AspNetCore.Mvc.Controller).</param>
+        /// <returns>A MemberAccessExpressionSyntaxNode. An IdentifierNameSyntax is returned when there is no concept of "left" and "right" nodes.</returns>
+        public static SyntaxNode BuildMemberAccessExpression(SyntaxGenerator generator, string qualifiedName)
+        {
+            if (generator is null)
+            {
+                throw new ArgumentNullException(nameof(generator));
+            }
+
+            if (qualifiedName is null || string.IsNullOrWhiteSpace(qualifiedName))
+            {
+                throw new ArgumentNullException(nameof(qualifiedName));
+            }
+
+            return BuildRecursiveNameSyntax(generator, qualifiedName.Split('.'), generator.MemberAccessExpression);
```

Comment:
Caching the arrays would only help if we're splitting the same strings repeatedly, but I can see how that could come up. I'll update it to cache the arrays.

Reply:
After further discussion, I'll leave this as is for now since it's not a high traffic code path. I'll add a comment, though, that we likely want to update this in the future.

Reply:
This split only occurs once during a code fixer action. Is the concern over about the caller being unaware of what this method does? If that's the case this overload could be removed and we would rely on the `BuildRecursiveNameSyntax` method.